### PR TITLE
Deal Client: decouple the user-facing interface (Solidity) from the SP-facing interface (CBOR).

### PR DESCRIPTION
**Pull request by raulk**
Also:

- Bring local types into this repo instead of relying on forks of filecoin-solidity.
- Move to using an npm-published version of filecoin-solidity.
- Adjust how extra params are represented.

## TODO

- Needs to be tested.
- Need to check with @ZakAyesh the feasiblity of using the higher-level types CommonTypes.Cid, CommonTypes.FilAddress, CommonTypes.BigInt in the `DealRequest` struct (as the hardhat kit may need helpers to produce these types).

### CODE REVIEW

**Comment by raulk** [comment]
* @nonsense is integrating into Boost.
* @nonsense is updating the datadaotool to work with this interface: https://github.com/nonsense/datadaotool/
* @ZakAyesh is working on hardhat tasks to call `makeDealProposal`, `getDealProposal`, `getExtraParams` from command line.
* @ssnin would love your eyes here.

File: contracts/basic-deal-client/DealClient.sol [hunk 1]
```diff
@@ -63,64 +98,72 @@ contract DealClient {
...
+        bytes32 _id = keccak256(abi.encodePacked(block.timestamp, msg.sender, index));
+        dealProposals[_id] = index;
 
-        ContractDealProposal memory proposal = deserializeContractDealProposal(deal);
-        pieceToProposal[proposal.piece_cid] = ProposalIdSet(_id, true);
+        pieceToProposal[deal.piece_cid.data] = ProposalIdSet(_id, true);
 
         // writes the proposal metadata to the event log
-        emit DealProposalCreate(_id, proposal.piece_size, proposal.verified_deal, 0); //bigIntToUint(proposal.storage_price_per_epoch));
+        emit DealProposalCreate(_id, deal.piece_size, deal.verified_deal, 0); //bigIntToUint(deal.storage_price_per_epoch));
```
Comment (shrenujbansal) [inline contracts/basic-deal-client/DealClient.sol:115]: change the last arg here to
`deal.storage_price_per_epoch`

File: contracts/basic-deal-client/DealClient.sol [hunk 2]
```diff
@@ -63,64 +98,72 @@ contract DealClient {
...
+        require(!pieceProviders[proposal.piece_cid.data].valid, "deal failed policy check: provider already claimed this cid");
 
+        pieceProviders[proposal.piece_cid.data] = ProviderSet(proposal.provider.data, true);
+        pieceDeals[proposal.piece_cid.data] = mdnp.dealId;
     }
 
     // client - filecoin address byte format
-    function addBalance(bytes memory client, uint256 value) public {
-
+    function addBalance(CommonTypes.FilAddress memory client, uint256 value) public {
```
Comment (shrenujbansal) [inline contracts/basic-deal-client/DealClient.sol:166]: Not critical for the demo or deal making flow but this API needs to remove the first arg as per the comments below. This was added for ease of testing before the market APIs were adapted

File: contracts/basic-deal-client/DealClient.sol [hunk 3]
```diff
@@ -131,28 +174,28 @@ contract DealClient {
...
     function bigIntToUint(CommonTypes.BigInt memory bigInt) internal view returns (uint256) {
-        BigNumber memory bigNumUint = BigNumbers.init(bigInt.val, bigInt.neg);
+        BigNumbers.BigNumber memory bigNumUint = BigNumbers.init(bigInt.val, bigInt.neg);
         uint256 bigNumExtractedUint = uint256(bytes32(bigNumUint.val));
         return bigNumExtractedUint;
     }
 
 
-    function withdrawBalance(bytes memory client, uint256 value) public returns(uint) {
+    function withdrawBalance(CommonTypes.FilAddress memory client, uint256 value) public returns(uint) {
```
Comment (shrenujbansal) [inline contracts/basic-deal-client/DealClient.sol:189]: same comment as addBalance

File: contracts/basic-deal-client/DealClient.sol [hunk 4]
```diff
@@ -39,20 +37,57 @@ struct ProviderSet {
     bool valid;
 }
 
-contract DealClient {
+// User request for this contract to make a deal. This structure is modelled after Filecoin's Deal
+// Proposal, but leaves out the provider, since any provider can pick up a deal broadcast by this
+// contract.
+struct DealRequest {
+    CommonTypes.Cid piece_cid;
```
Comment (nonsense) [inline contracts/basic-deal-client/DealClient.sol:44]: I suggest we change this to:
```suggestion
   bytes piece_cid;
```

File: contracts/basic-deal-client/DealClient.sol [hunk 5]
```diff
@@ -39,20 +37,57 @@ struct ProviderSet {
...
+    uint64 piece_size;
+    bool verified_deal;
+    CommonTypes.FilAddress client;
```
[the remark this answers is not in the document]
Comment (raulk) [reply contracts/basic-deal-client/DealClient.sol]: Agree, fixed.

File: contracts/basic-deal-client/DealClient.sol (hunk 4 above)
Comment (raulk) [reply contracts/basic-deal-client/DealClient.sol]: Agree, fixed.

File: contracts/basic-deal-client/DealClient.sol (hunk 1 above)
Comment (raulk) [reply contracts/basic-deal-client/DealClient.sol]: Done!

**Pull request merged**